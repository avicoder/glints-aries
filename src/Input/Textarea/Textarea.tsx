import React, { FC, HTMLAttributes, RefObject } from 'react';

import classNames from 'classnames';
import isFunction from 'lodash/isFunction';

import {
  TextareaContainer,
  TextareaInput,
  TextareaLabel,
} from './TextareaStyle';

const MIN_ROWS = 4;
const MAX_ROWS = 12;

/** Use <code>onChange</code> to listen to input changes. Use
 * <code>onBlur</code> to listen to focus loss. */
export const Textarea: FC<Props> = ({
  label,
  value = undefined,
  status = 'default',
  disabled = false,
  className,
  onBlur,
  onChange,
  forwardedRef,
  removeFloatingLabel = false,
  small,
  ...restProps
}) => {
  const [floating, setFloating] = React.useState<boolean>(false);
  const [rows, setRows] = React.useState<number>(MIN_ROWS);
  const innerRef = React.useRef(null);
  const textareaInputRef = forwardedRef || innerRef;

  const [textareaMaxHeight, setTextareaMaxHeight] = React.useState<number>(0);

  React.useLayoutEffect(() => {
    const textarea = textareaInputRef.current;

    if (!textarea) return;

    setTextareaMaxHeight(~~(textarea.offsetHeight * 3.7 + 23 * 8 + 1));
  }, [textareaInputRef]);

  const handleChange = React.useCallback(
    (e: React.ChangeEvent<HTMLTextAreaElement>) => {
      const previousRows = e.target.rows;
      e.target.rows = MIN_ROWS;

      const currentRows = ~~(e.target.scrollHeight / 30);

      if (currentRows === previousRows) {
        e.target.rows = currentRows;
      }

      if (currentRows >= MAX_ROWS) {
        e.target.rows = MAX_ROWS;
        e.target.scrollTop = e.target.scrollHeight;
      }

      setRows(currentRows < MAX_ROWS ? currentRows : MAX_ROWS);

      if (isFunction(onChange)) {
        return onChange(e);
      }
    },
    [onChange]
  );

  const handleFocusChange = React.useCallback(
    (e: React.FocusEvent<HTMLTextAreaElement>) => {
      setFloating(e.target.value.length > 0);

      if (onBlur !== undefined) {
        return onBlur(e);
      }
    },
    [onBlur]
  );

  React.useEffect(() => {
    if (value && value !== '' && !floating) {
      setFloating(true);
    }
  }, [value, floating]);

  return (
    <TextareaContainer className={classNames('aries-textarea', className)}>
      <TextareaInput
        ref={textareaInputRef}
        placeholder={removeFloatingLabel && label}
        rows={rows}
        status={status}
        disabled={disabled}
        onBlur={handleFocusChange}
        onChange={handleChange}
        floating={floating}
        value={value}
        aria-label={label}
        small={small}
        {...restProps}
        style={{
          maxHeight: `${textareaMaxHeight}px`,
        }}
      />
      {!removeFloatingLabel && (
        <TextareaLabel
          data-testid="textarea-label"
          floating={floating}
          status={status}
          small={small}
        >
          {label}
        </TextareaLabel>
      )}
    </TextareaContainer>
  );
};

export type Status = 'default' | 'error';

export interface Props extends HTMLAttributes<HTMLTextAreaElement> {
  value?: string;
  status?: Status;
  /** Placeholder for the text area. */
  label?: string;
  removeFloatingLabel?: boolean;
  disabled?: boolean;
  forwardedRef?: RefObject<HTMLTextAreaElement>;
}

const forwardRef = (props: Props, ref: RefObject<HTMLTextAreaElement>) => (
  <Textarea {...props} forwardedRef={ref} />
);

forwardRef.displayName = Textarea.name;

export default React.forwardRef(forwardRef);
