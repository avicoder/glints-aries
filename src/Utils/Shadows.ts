export const Shadows = {
  1: {
    up: '0 2px 4px 0 rgba(0, 0, 0, 0.15)',
  },
  2: {
    up: '0 4px 8px 0 rgba(0, 0, 0, 0.15)',
  },
  3: {
    up: '0 6px 12px 0 rgba(0, 0, 0, 0.10)',
  },
  4: {
    up: '0 12px 24px 0 rgba(0, 0, 0, 0.10)',
    down: '0 -6px 12px 0 rgba(0, 0, 0, 0.10)',
  },
};
