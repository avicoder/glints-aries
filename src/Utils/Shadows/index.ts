const ShadowOneDown = '0 2px 4px 0 rgba(0, 0, 0, 0.15)';
const ShadowTwoDown = '0 4px 8px 0 rgba(0, 0, 0, 0.15)';
const ShadowThreeDown = '0 6px 12px 0 rgba(0, 0, 0, 0.10)';
const ShadowFiveDown = '0 12px 24px 0 rgba(0, 0, 0, 0.10)';
const ShadowFiveUp = '0 -6px 12px 0 rgba(0, 0, 0, 0.10)';

export const Shadows = {
  1: {
    down: ShadowOneDown,
  },
  2: {
    down: ShadowTwoDown,
  },
  3: {
    down: ShadowThreeDown,
  },
  4: {
    down: ShadowFiveDown,
    up: ShadowFiveUp,
  },
} as const;

export type ShadowValueType =
  | typeof ShadowOneDown
  | typeof ShadowTwoDown
  | typeof ShadowThreeDown
  | typeof ShadowFiveDown
  | typeof ShadowFiveUp;
