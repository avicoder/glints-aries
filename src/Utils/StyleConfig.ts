export enum ETabAlignment {
  HORIZONTAL = 'horizontal',
  VERTICAL = 'vertical',
}

export enum ETabThemeVariant {
  BLUE = 'blue',
  BLACK = 'black',
}

export enum ETooltipPosition {
  TOP = 'top',
  BOTTOM = 'bottom',
  LEFT = 'left',
  RIGHT = 'right',
}

export const ButtonVariant = {
  SOLID_WHITE: 'solid-white',
  SOLID_BLUE: 'solid-blue',
  YELLOW: 'yellow',
  GHOST: 'ghost',
  WHITE_GREY: 'white-grey',
  LINK: 'link',
};

export const RoundedCornerButtonVariant = {
  SOLID_WHITE: 'solid-white',
  SOLID_BLUE: 'solid-blue',
  GHOST: 'ghost',
  WHITE_GREY: 'white-grey',
};

export const PsychedelicTheme = {
  BLUE_DOMINANT: 'blue-dominant',
  BLUE_DEFAULT: 'blue-default',
  RED_DOMINANT: 'red-dominant',
  RED_DEFAULT: 'red-default',
  YELLOW_DOMINANT: 'yellow-dominant',
  YELLOW_DEFAULT: 'yellow-default',
};

export const ButtonTheme = {
  BLUE: 'blue',
  WHITE: 'white',
};

export const ScreenSize = {
  mobileS: 320,
  mobileM: 480,
  mobileL: 640,
  tablet: 768,
  desktopS: 1024,
  desktopM: 1260,
  desktopL: 1440,
};

export const Device = {
  mobileS: `(max-width: ${ScreenSize.mobileS}px)`,
  mobileM: `(max-width: ${ScreenSize.mobileM}px)`,
  mobileL: `(max-width: ${ScreenSize.mobileL}px)`,
  tablet: `(max-width: ${ScreenSize.tablet}px)`,
  desktopS: `(max-width: ${ScreenSize.desktopS}px)`,
  desktopM: `(max-width: ${ScreenSize.desktopM}px)`,
  desktopL: `(max-width: ${ScreenSize.desktopL}px)`,
  desktopXL: `(min-width: ${ScreenSize.desktopL + 1}px)`,
};
