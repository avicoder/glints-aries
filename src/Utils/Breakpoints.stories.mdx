import { Meta } from '@storybook/addon-docs/blocks';
import { ScreenSize, Device } from './StyleConfig';

<Meta title="Utilities/Breakpoints" />

# Breakpoints

These utilities are used internally by components like
[GlintsContainer](/?path=/docs/utilities-glintscontainer--page#glintscontainer).
You should use them when building your own responsive styles to ensure that
breaks only happen at a few well-defined viewport-widths.

## Device

Our device classes. Names are generally oriented upwards: For example, when you
use `Device.desktopS`, your styles will apply for all viewport-widths greater or
equal to `1024px`. That means that you have to override those styles with rules
for wider viewports if needed. Thus, if you have styles for serveral widths, the
rules for narrow viewports should be at the top and should be consecutively
overridden by rules for wider viewports towards the bottom of your css.

`mobileS` is an exception: It applies "downwards", i.e. on all screens smaller
than `480px`.

```TypeScript
import { Device } from 'glints-aries';

const MyContainer = styled.div`
  // Applies on screens smaller than tablet
  color: red;

  // Applies from tablets to screens smaller than large desktops
  @media ${Device.tablet} {
    color: green;
  }

  // Applies on large desktops
  @media ${Device.desktopL} {
    color: blue;
  }
`;
```

<table>
  <thead>
    <tr>
      <th>Device</th>
      <th>Value</th>
    </tr>
  </thead>
  <tbody>
    {Object.keys(Device).map(dev => (
      <>
        <tr>
          <td>{dev}</td>
          <td>{Device[dev]}</td>
        </tr>
      </>
    ))}
  </tbody>
</table>

## ScreenSize

These are the numeric pixel values that define our breakpoints. The values are
the smallest pixel width that should be classified as the respective constant's
size class. For example, a viewport needs to be at least `1440px` wide to "qualify" as
Desktop Large. Note that the values do not include the `px` prefix;

```TypeScript
import { ScreenSize } from 'glints-aries';

const MyContainer = styled.div`
  color: red;
  @media (min-width: ${ScreenSize.desktopL}px) {
    color: blue;
  }
`;
```

<table>
  <thead>
    <tr>
      <th>Device</th>
      <th>Value</th>
    </tr>
  </thead>
  <tbody>
    {Object.keys(ScreenSize).map(size => (
      <>
        <tr>
          <td>{size}</td>
          <td>{ScreenSize[size]}</td>
        </tr>
      </>
    ))}
  </tbody>
</table>
