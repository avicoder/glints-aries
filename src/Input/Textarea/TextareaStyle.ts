import styled from 'styled-components';
import { Greyscale, PrimaryColor, SecondaryColor } from '../../Utils/Colors';

export const TextareaContainer = styled.div`
  flex-grow: 1;
  position: relative;
  display: flex;
`;

export const TextareaLabel = styled.label<TextareaLabelProps>`
  position: absolute;
  left: 22px;
  top: 1em;
  background: ${Greyscale.white};
  color: ${({ floating }) =>
    floating ? `${Greyscale.black}` : `${Greyscale.grey}`};
  transition: all 0.2s;
  pointer-events: none;
  font-weight: 300;
  font-size: 1.1em;

  ${({ floating }) => {
    if (floating) {
      return `
        padding: 0 5px;
        transform: translate3d(-15px,-20px,0);
        transition: all .2s;
        font-size: 12px;
      `;
    }
  }}

  ${({ status, floating }) => {
    if (status === 'error' && floating) {
      return `
        color: ${PrimaryColor.glintsred};
      `;
    }
  }}
`;

interface TextareaLabelProps {
  floating: boolean;
  status?: string;
}

export const TextareaInput = styled.textarea<TextareaInputProps>`
  position: relative;
  width: 100%;
  outline: none;
  font-size: 1.1em;
  line-height: 1.5;
  padding: 15px 20px;
  border: ${({ status }) =>
    status === 'error'
      ? `2px solid ${PrimaryColor.glintsred}`
      : `2px solid ${Greyscale.grey}`};
  transition: border .5s;
  resize: vertical;
  overflow: auto;
  height: auto;

  ${({ status, floating }) => {
    if (status === 'error') {
      if (floating) {
        return `
          border: 2px solid ${PrimaryColor.glintsred};
        `;
      }
    }
  }}
    

  &:disabled {
    cursor: not-allowed;
    background: ${Greyscale.softgrey};

    + ${TextareaLabel} {
      background: transparent;
      color: ${Greyscale.grey};
    }

    &:hover {
      border: 2px solid ${Greyscale.lightgrey};
    }
  }

  &:hover {
    border: 2px solid ${SecondaryColor.actionblue};
    transition: border .5s;

    ${({ status }) => {
      if (status === 'error') {
        return `
          border: 2px solid ${PrimaryColor.glintsred};
        `;
      }
    }}

    + ${TextareaLabel} {
      color: ${Greyscale.black};

      ${({ status }) => {
        if (status === 'error') {
          return `
          color: ${PrimaryColor.glintsred};
        `;
        }
      }}
    }
  }

  &:focus {
    border: 2px solid ${SecondaryColor.actionblue};

    ${({ status }) => {
      if (status === 'error') {
        return `
          border: 2px solid ${PrimaryColor.glintsred};
        `;
      }
    }}

    + ${TextareaLabel} {
      padding: 0 5px;
      transform: translate3d(-15px,-20px,0);
      transition: all .2s;
      color: ${Greyscale.black};
      font-size: 12px;

      ${({ status }) => {
        if (status === 'error') {
          return `
          color: ${PrimaryColor.glintsred};
        `;
        }
      }}
    }
  }
`;

interface TextareaInputProps {
  floating?: boolean;
  status?: string;
}
