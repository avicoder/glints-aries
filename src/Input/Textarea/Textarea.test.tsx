import * as React from 'react';

import * as renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Textarea from './Textarea';
import { SecondaryColor, PrimaryColor } from '../../Utils/Colors';

interface Props {
  label: string;
  value: string;
  onChange: jest.Mock;
}

const props: Props = {
  label: 'Description',
  value: 'Explanation',
  onChange: jest.fn().mockImplementation(event => event.target.value),
};

const inputValue =
  'As the first sign in the zodiac, the presence of Aries always marks the beginning of something energetic and turbulent.';

function setupTextarea(props: Props, otherProps?: any) {
  const { getByLabelText, ...utils } = render(
    <Textarea label={props.label} onChange={props.onChange} {...otherProps} />
  );
  const textareaInput = getByLabelText(props.label) as HTMLTextAreaElement;
  return { textareaInput, utils };
}

function setupTextareaWithChangeEvent(props: Props) {
  const { textareaInput } = setupTextarea(props);
  userEvent.type(textareaInput, inputValue);
  return { textareaInput };
}

function createNodeMock(element: any) {
  if (element.type === 'textarea') {
    return {
      value: 0,
      offsetHeight: 34,
    };
  }
  return null;
}

it('<Textarea> should render with a value and label of Description', () => {
  const options = { createNodeMock };
  const TextareaSnapshot = renderer
    .create(<Textarea label={props.label} value={props.value} />, options)
    .toJSON();
  expect(TextareaSnapshot).toMatchSnapshot();
});

describe('when it is rendered', () => {
  it('should display the correct input label', () => {
    const { textareaInput } = setupTextarea(props);
    expect(textareaInput).toBeTruthy();
  });

  it('should have an empty string as its value', () => {
    const { textareaInput } = setupTextarea(props);
    expect(textareaInput.value).toEqual('');
  });
});

describe('when a value is entered', () => {
  it('should call onChange for each input', () => {
    setupTextareaWithChangeEvent(props);
    expect(props.onChange).toHaveBeenCalledTimes(inputValue.length);
  });

  it('should return the input value when onChange is passed a function: event => event.target.value', () => {
    setupTextareaWithChangeEvent(props);
    expect(props.onChange.mock.results.slice(-1)[0].value).toEqual(inputValue);
  });

  it('should display the correct value', () => {
    const { textareaInput } = setupTextareaWithChangeEvent(props);
    expect(textareaInput.value).toEqual(inputValue);
  });
});

describe('when status is:', () => {
  it("'success', it should display a lightblack border", () => {
    const { textareaInput } = setupTextarea(props, { status: 'success' });
    expect(textareaInput).toHaveStyle(`
      border-color: ${SecondaryColor.lightblack};
    `);
  });

  it("'error', it should display a red border", () => {
    const { textareaInput } = setupTextarea(props, { status: 'error' });
    expect(textareaInput).toHaveStyle(`
      border-color: ${PrimaryColor.glintsred};
    `);
  });
});

describe('when removeFloatingLabel is true', () => {
  it('should remove the floating label', () => {
    const { textareaInput } = setupTextarea(props, {
      removeFloatingLabel: true,
    });
    const textFieldLabel = textareaInput.parentElement.querySelector(
      '[data-testid="textarea-label"]'
    ) as HTMLLabelElement;
    expect(textareaInput.parentElement).not.toContainElement(textFieldLabel);
  });

  it('should display a placeholder', () => {
    const { utils } = setupTextarea(props, { removeFloatingLabel: true });
    const textareaInput = utils.getByPlaceholderText(props.label);
    expect(textareaInput).toBeTruthy();
  });
});

describe('<Textarea /> forwards ref to underlying textarea element', () => {
  const label = 'forward-ref';

  test('ref is being forwarded correctly', () => {
    const ref = React.createRef<HTMLTextAreaElement>();
    const { queryByLabelText } = render(
      <Textarea ref={ref} label={label} value={props.value} />
    );
    const textarea = queryByLabelText(label);
    expect(textarea).toEqual(ref.current);
  });

  test('underlying textarea can be focused/blurred through ref', () => {
    const ref = React.createRef<HTMLTextAreaElement>();
    const fn = jest.fn();
    const { queryByLabelText } = render(
      <Textarea ref={ref} label={label} onBlur={fn} />
    );
    const textarea = queryByLabelText(label);
    expect(textarea).not.toHaveFocus();

    ref.current.focus();
    expect(textarea).toHaveFocus();

    ref.current.blur();
    expect(textarea).not.toHaveFocus();
  });
});
