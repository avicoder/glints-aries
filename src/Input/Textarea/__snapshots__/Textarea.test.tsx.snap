// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Textarea> should render with a label of Description 1`] = `
<div
  className="TextareaStyle__TextareaContainer-sc-1998n74-0 kjIlzy aries-textarea"
>
  <textarea
    aria-label="Description"
    className="TextareaStyle__TextareaInput-sc-1998n74-2 kjACxw"
    onBlur={[Function]}
    onChange={[Function]}
    rows={4}
    style={
      Object {
        "maxHeight": "310px",
      }
    }
  />
  <label
    className="TextareaStyle__TextareaLabel-sc-1998n74-1 iCsNCu"
    data-testid="textarea-label"
  >
    Description
  </label>
</div>
`;
